source/common/upstream/upstream_impl.cc

```diff
@@ -624,20 +625,40 @@ void HostSetImpl::rebuildLocalityScheduler(
   locality_scheduler = nullptr;
   if (all_hosts_per_locality != nullptr && locality_weights != nullptr &&
       !locality_weights->empty() && !eligible_hosts.empty()) {
-    locality_scheduler = std::make_unique<EdfScheduler<LocalityEntry>>();
-    locality_entries.clear();
-    for (uint32_t i = 0; i < all_hosts_per_locality->get().size(); ++i) {
-      const double effective_weight = effectiveLocalityWeight(
-          i, eligible_hosts_per_locality, *excluded_hosts_per_locality, *all_hosts_per_locality,
-          *locality_weights, overprovisioning_factor);
-      if (effective_weight > 0) {
-        locality_entries.emplace_back(std::make_shared<LocalityEntry>(i, effective_weight));
-        locality_scheduler->add(effective_weight, locality_entries.back());
+    if (Runtime::runtimeFeatureEnabled(
+            "envoy.reloadable_features.edf_lb_locality_scheduler_init_fix")) {
+      locality_entries.clear();
+      for (uint32_t i = 0; i < all_hosts_per_locality->get().size(); ++i) {
+        const double effective_weight = effectiveLocalityWeight(
+            i, eligible_hosts_per_locality, *excluded_hosts_per_locality, *all_hosts_per_locality,
+            *locality_weights, overprovisioning_factor);
+        if (effective_weight > 0) {
+          locality_entries.emplace_back(std::make_shared<LocalityEntry>(i, effective_weight));
+        }
+      }
+      // If not all effective weights were zero, create the scheduler.
+      if (!locality_entries.empty()) {
+        locality_scheduler = std::make_unique<EdfScheduler<LocalityEntry>>(
+            EdfScheduler<LocalityEntry>::createWithPicks(
```

Comment:
This is LGTM overall. 

I have only one question. The `createWithPicks` will use the `seed % 429496729` as the pre picks number. 
And because the seed is a random value from `random()` method, so it's possible we may do  429496729 times pre-picks when we update the hosts.

Do we acutally evalute the possible performance impact at the worst case? If the users update the host set frequently, will this bring huge performance burden?

(PS: the host scheduler will use `seed % hosts.size()` as the pre picks number, I think this should so some similar things? like we can use `seed % locality_entries.size()` or `std::max(8, seed % locality_entries.size())` here?

Reply:
> This is LGTM overall.
> 
> I have only one question. The `createWithPicks` will use the `seed % 429496729` as the pre picks number. And because the seed is a random value from `random()` method, so it's possible we may do 429496728 times pre-picks when we update the hosts.
> 
> Do we acutally evalute the possible performance impact at the worst case? If the users update the host set frequently, will this bring huge performance burden?
> 

While conceptually the number of pre-picks could be large, the code is bounded by the number of entries/weights (N- see [here](https://github.com/envoyproxy/envoy/blob/20c7368afa9d686a109f9601ae1b9b6028b74b0a/source/common/upstream/edf_scheduler.h#L136)), and the operation has the complexity of O(N*log(N)), which is the same as it is right now (because the localities are added one after the other to the priority-queue).
A proof of the bounded number of iterations can be found at the doc linked for [this comment](https://github.com/envoyproxy/envoy/pull/31592#issue-2063014213).

> (PS: the host scheduler will use `seed % hosts.size()` as the pre picks number, I think this should be some similar things? like we can use `seed % locality_entries.size()` or `std::max(8, seed % locality_entries.size())` here?

Yes, the next PR will be to fix the wrong initialization there as well.

Reply:
I see. I actually not completely get it. I will read the code more detailedly. Will unblock the PR first.

Reply:
feel free to reach out directly and I can try to give a guided explanation.

Reply:
I got it. Sorry for the previous negligenc.

I have to say, it's awesome 👍.